development_guidelines.md

```diff
@@ -0,0 +1,415 @@
+as many checks as possible into the continuous integration process.
+However, some things are currently manual or simply cannot be automated.  
+In general the reviewer has to check/review the following:
+- Specification:
+  - The feature and its changes to the system are comprehensively specified.
+  - All requirements are exhaustively tested by one or more tests.
+  - All specifications are linked against their corresponding tests 
+(how-to link see [here](#how-to-link-a-requirement-against-a-test))
+- User interface:
+  - The user interface is easy to understand and use.
+  - The user interface is comprehensively documented.
+  - If an existing user interface is broken by the new feature, then the
+interface version has to be adapted according the nature of the change.
+- System architecture:
+  - The system architecture complies with the clean architecture guidelines.
+(see "Clean Architecture" by Robert C. Martin)
+  - The architecture diagrams are up to date and easily understandable.
+- Code:
+  - The code has to comply with the clean code rules 
+(see "Clean Code: A Handbook of Agile Software Craftsmanship" by Robert C. Martin).
+- Tests:
+  - The tests are easy to understand (e.g. each test only checks one aspect 
+of a component, etc.)
+  - The tests are comprehensively documented (how-to document tests see
+[here](#how-to-document-tests))
+
+
+## Continuous integration (CI)
+All pull requests (PR's) have to be approved by the continuous integration tool
+[Travis](https://travis-ci.org/PilzDE/).  
+
+### How-to activate Travis for forked repositories
+If you are an external developer and want to create a pull request, you can
+do so by forking one of our repositories, make changes, and then create a pull
+request on our repository.  
+To check that the continuous integration runs without problems on 
+*your* forked repository, you have to activate Travis for your repository.
+For more information on how to do that have a look at the
+[Travis tutorials](https://docs.travis-ci.com/user/tutorial/#to-get-started-with-travis-ci).  
+  
+**Note:**  
+In our repositories already exists a `.travis.yml` file, so there is no need
+to create your own.
+
+### How-to run the CI processes locally
+If you have a look at our
+[Travis file](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/.travis.yml#L34), 
+you see that our CI is based on the 
+[ROS industrial CI configurations](https://github.com/ros-industrial/industrial_ci)
+for ROS. In some cases, for example in case of errors only appearing on 
+Travis, etc., it might be advantageous to do *exactly* what Travis does locally 
+(At least in terms of building the packages and executing the tests).
+To achieve this, have a look at the explanations given in the 
+[ros-industrial repository](https://github.com/ros-industrial/industrial_ci/blob/legacy/doc/index.rst#run-industrial_ci-on-local-host).
+
+## Code style guide
+<!---TODO: Are we really compliant with the ROS coding style guide? --->
```

Comment:
Are we?

Reply:
Don't know. But at one time we said we want too. I simply deleted the TODO :) -> Done.